source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -194,7 +194,7 @@ void DnsResolverImpl::AddrInfoPendingResolution::onAresGetAddrInfoCallback(
     //
     // The channel cannot be destroyed and reinitialized here because that leads to a c-ares
     // segfault.
-    if (status == ARES_ECONNREFUSED) {
+    if (status == ARES_ECONNREFUSED || status == ARES_EREFUSED) {
```

Comment:
c-ares used to report 3 responses as ARES_ECONNREFUSED, which now get bubbled up as the original response codes. https://github.com/c-ares/c-ares/blob/ebb0a631ef10c88478e0da4eb5176807bfa0ef54/src/lib/ares_process.c#L711

I think this needs to check for all 3 status now: ARES_ESERVFAIL, ARES_ENOTIMP and ARES_EREFUSED